### PR TITLE
providers/aliyun: Add aliyun

### PR DESCRIPTION
**Pull request by ashcrow**
Not 100% sure if this would be enough, but it's a start :smile:

See:
- https://www.alibabacloud.com/help/doc-detail/49122.htm
- https://github.com/coreos/ignition/issues/589

### CODE REVIEW

**Comment by lucab** [comment]
ok to test

File: internal/providers/aliyun/aliyun.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2016 CoreOS, Inc.
```
Comment (lucab) [inline internal/providers/aliyun/aliyun.go]: `2019 Red Hat`

**Comment by ajeddeloh** [comment]
LGTM if it works, haven't tested

**Comment by lucab** [comment]
As a side-note, I just did a manual check and the `user-data` endpoint directly returns a 404 if the no config was provide when creating the instance. @ajeddeloh how does/should Ignition react to that?

**Comment by ashcrow** [comment]
It looks like [`fetchToBuffer`](https://github.com/coreos/ignition/blob/master/internal/resource/url.go#L106-L123) would end up passing back the error from the [underlying http call](https://github.com/coreos/ignition/blob/master/internal/resource/url.go#L265-L266).

Since it's valid to have no user-data at all we could catch the 404 response in the provider and pass an empty JSON struct into `util.ParseConfig`.

Thoughts?

**Comment by ashcrow** [comment]
Something like:

```diff
diff --git a/internal/providers/aliyun/aliyun.go b/internal/providers/aliyun/aliyun.go
index 7b8a879..90858ff 100644
--- a/internal/providers/aliyun/aliyun.go
+++ b/internal/providers/aliyun/aliyun.go
@@ -39,6 +39,16 @@ func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
        data, err := f.FetchToBuffer(userdataUrl, resource.FetchOptions{
                Headers: resource.ConfigHeaders,
        })
+       // Aliyun returns a 404 if no user-data is set. If this is the case, intercept
+       // the response, re-try the error, and set the data to be empty
+       // See: https://github.com/coreos/ignition/pull/869
+       if err == resource.ErrNotFound {
+               f.Logger.Debug("setting user-data empty as no user-data was set in or provided by the endpoint")
+               data.Reset()
+               // If we can not write the empty JSON to the buffer, capture the error
+               // and error out of FetchConfig
+               _, err = data.WriteString("{}")
+       }
        if err != nil {
                return types.Config{}, report.Report{}, err
        }
```

**Comment by ajeddeloh** [comment]
If not providing a config gives a 404, we should treat it the same as if no config was provided, so catching the 404 sgtm.

@ashcrow We actually want to return an error here, with an empty `Config{}`. This will actually do that since `{}` is an invalid config and `Parse()` will choke, but I don't think that's what you intended and returning an explicit error would be more clear.

If a provider returns an error that's ok, Ignition will treat that the same as no config. Ignition actually tries the cmdline provider (`ignition.config.url` kcmdline), then the installed config (the one coreos-installer writes) then the platform specific one. If those encounter errors, that's fine, they just log it and fall back to the next one.

**Comment by ashcrow** [comment]
@ajeddeloh Awesome, perfect!

**Comment by arithx** [comment]
@ashcrow: I re-ran the specific subjob, was just a travis flake.

**Comment by ajeddeloh** [comment]
LGTM, this has been tested in cloud with an empty config, yes?

**Comment by ashcrow** [comment]
Verifying now @ajeddeloh, almost done :-)

**Comment by ashcrow** [comment]
@ajeddeloh Verified ssh access and file/dir writing with this change (thanks to @arithx for providing an image to use!). It worked great!

**Comment by ashcrow** [comment]
Everything BUT an empty config passes. @arithx pointed out https://github.com/coreos/ignition/blob/master/internal/providers/aws/aws.go#L47 for 404 responses. I'll see about doing that as well here.

**Comment by ashcrow** [comment]
> Everything BUT an empty config passes. @arithx pointed out https://github.com/coreos/ignition/blob/master/internal/providers/aws/aws.go#L47 for 404 responses. I'll see about doing that as well here.

Updated to be inline with the above.

**Comment by ashcrow** [comment]
This now works with and without a config provided in Aliyuns metadata.

**Pull request merged**